Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h

```diff
@@ -28,25 +28,64 @@
 #include "GPUObjectDescriptorBase.h"
 #include "GPUPredefinedColorSpace.h"
 #include "HTMLVideoElement.h"
+#include "WebCodecsVideoFrame.h"
 #include <pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h>
 #include <wtf/RefPtr.h>
 
+typedef struct __CVBuffer* CVPixelBufferRef;
+
 namespace WebCore {
 
 class HTMLVideoElement;
+#if ENABLE(WEB_CODECS)
+using GPUVideoSource = std::variant<RefPtr<HTMLVideoElement>, RefPtr<WebCodecsVideoFrame>>;
+#else
+using GPUVideoSource = RefPtr<HTMLVideoElement>;
+#endif
 
 struct GPUExternalTextureDescriptor : public GPUObjectDescriptorBase {
+
+    static PAL::WebGPU::VideoSourceIdentifier mediaIdentifierForSource(const GPUVideoSource& videoSource, CVPixelBufferRef& outPixelBuffer)
+    {
+#if ENABLE(WEB_CODECS)
+        return WTF::switchOn(videoSource, [] (const RefPtr<HTMLVideoElement> videoElement) -> PAL::WebGPU::VideoSourceIdentifier {
+            auto playerIdentifier = videoElement->playerIdentifier();
+            return PAL::WebGPU::HTMLVideoElementIdentifier { playerIdentifier ? playerIdentifier->toUInt64() : 0 };
+        }
+        , [&outPixelBuffer] (const RefPtr<WebCodecsVideoFrame> videoFrame) -> PAL::WebGPU::VideoSourceIdentifier {
+#if PLATFORM(COCOA)
```

Comment:
Why is it COCOA specific?

Reply:
Oh it doesn't need to be COCOA specific, only because it is calling `pixelBuffer`. But I think I can remove this by passing an opaque pointer through the PAL layer and working with VideoFrame instances directly.